app/models/task.rb

```diff
@@ -45,20 +45,40 @@ class Task < ApplicationRecord
   scope :min_stars, ->(stars) { average_rating.where('COALESCE(avg_rating, 0) >= ?', stars) }
   scope :sort_by_average_rating_asc, -> { average_rating.order(average_rating: 'ASC') }
   scope :sort_by_average_rating_desc, -> { average_rating.order(average_rating: 'DESC') }
+  scope :access_level, ->(access_level) { where(access_level:) }
 
   serialize :meta_data, HashAsJsonbSerializer
 
+  enum access_level: {private: 0, public: 1}, _default: :private, _prefix: true
+
   def self.ransackable_scopes(_auth_object = nil)
-    %i[created_before_days min_stars]
+    %i[created_before_days min_stars access_level]
   end
 
   def self.ransackable_attributes(_auth_object = nil)
     %w[title description programming_language_id created_at]
   end
 
   def can_access(user)
+    author?(user)
+  end
+
+  def author?(user)
     self.user == user
   end
+  # TODO: Find a better name for the methods
+
+  def in_same_group?(user)
+    in_same_group_member?(user) || in_same_group_admin?(user)
+  end
+
+  def in_same_group_member?(user)
+    groups.any? {|group| group.confirmed_member?(user) }
+  end
+
+  def in_same_group_admin?(user)
+    groups.any? {|group| group.admin?(user) }
+  end
```

Comment:
I'm not really happy with these methods as well, but have no idea how to fix this currently, will come back to this

Reply:
If you have an idea, I would really appreciate that. Otherwise, we just leave it for now.

Reply:
it seems counter-intuitive to solve it like this, but I can't think of a better way, I think we can leave it as well